### PR TITLE
Improve the memory usage of the async destination

### PR DESCRIPTION
**Pull request by benmoriceau**
## What
The Async framework is facing some memory management issue which makes the amount of memory we believe that we use to be smaller than what we actually use. In addition to that we were never removing the store state from the memory 

## Recommended reading order
1. `AsyncStreamConsumer.java` => Make sure that we remove the JsonNode after storing the serialized version in the serialized field
2. `GlobalAsyncStateManager.java` => Clean up state from the in memory storage after emitting it.
3. `BufferManager.java` => Allow logging to the memory used by the state.
4. Tests

Related to: https://github.com/airbytehq/oncall/issues/2957

### CODE REVIEW

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java
Comment (jbfbell) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java:308]: I feel pretty strongly we should not be doing this - this is an expectation of what the memory should be but does not conform to what the JVM is doing with memory. In practice this means engineers debugging connections might see this statement and think its true when its an estimate. I would recommend at least one of the following changes:

- remove this method, and its call from `BufferManager` (at the very least, don't repeat the pattern) and remove lines 79-85 in `BufferManager`
- remove the calculated values (`memoryAllocated`, `memoryUsed`, maybe all the values from `GlobalMemoryManager` class and use the `Runtime.getRuntime().totalMemory()` and `Runtime.getRuntime().freeMemory()` instead

**Review by jbfbell — CHANGES_REQUESTED** [review]

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java
Comment (benmoriceau) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java:308]: I understand that it is confusing but the goal of this log is to give an estimation of what we are storing in the buffer and not the exact value of what the memory usage is. While debugging we used VisualVM in order to access to the total and free memory.
This estimation was wrong because of what this PR is aiming to fix.

Comment (jbfbell) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java:308]: I am missing something - I think I don't understand why its useful to log an estimate? Especially when there are better resources available for memory usage? My concern is that this gives a false impression of reality. In a vacuum with no context of your project, I would (and did) assume this was accurately reporting memory usage. At the very least, this log line should be changed to be explicit that it does not reflect actual measurements. Something like `bytes expected to be written (not an accurate representation of memory usage)` thats maybe too verbose but I want to be clear what my objection is here.

Comment (tryangul) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java:308]: > I would (and did) assume this was accurately reporting memory usage.

We thought the same thing when we inherited this! Totally agree it's not ideal, however, after being in the code for a while, I think I understand the general context it was written in.

The goal, afaict, is to provide a bounded topic of queues (partitioned by stream), such that as we read messages from upstream we can appropriately backpressure before we run out of memory. The catch is we don't know the size of the messages we are enqueuing until runtime, so we can't use static array sizes or what have you. The GlobalMemoryManager and its queues are an attempt to provide this functionality, but they have some issues.
 
These are 
1) misleading naming
2) poor sizing of the available memory—it is way too large so it never backpressures (we probably need to size per dest)
3) the estimate size of the messages is off (we are trying to improve this)
4) general bugs

Our tack thus far has been to incrementally improve it, so the log messages have been useful in that regard.

Comment (jbfbell) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java:308]: ok if the messages have some use, the minimum change here is to the text in the log lines. Making them explicitly clear what their purpose is so that there can be no confusion. To be blunt, in their current state these log lines are more harmful than good to someone not on this project.

Comment (benmoriceau) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java:308]: Updated

Comment (tryangul) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java:308]: I'd also argue we should rename the class / methods as well. Maybe in a follow-up PR to unblock @benmoriceau

**Review by jbfbell — APPROVED** [review]

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java
Comment (davinchia) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java:147]: note: I'm pretty sure iterating over a concurrent hash map is thread safe as the iterator returns a snapshot of the map when it's first created.

the synchronized block was originally scoped to the `stateIdQ` since this is a block of operations we need to do as a transaction across a bunch of queues.

this code is fine as is. just sharing some background context.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java
Comment (davinchia) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java:167]: I think this comment should be on line 160. 

My understanding: we can add a stateId that doesn't have an associated state object yet.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java
Comment (davinchia) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java:169]: I think this is only true if we make sure to add a counter before we add the state id to the queue - I think we can do this if line 334 comes before line 333.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java
Comment (davinchia) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java:137]: nit: `final`

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java
Comment (davinchia) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java:144]: nit: what about adding a comment here explaining we want this to stay as String to relieve memory pressure, and it's less complex to leverage Jackson to unmarshall the json and later drop it, rather than write a custom deserialiser?

**Review by davinchia — COMMENTED** [review]
Thanks!

Couple of comments to resolve in GlobalAsyncStateManager before approving. Otherwise looks good!

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java
Comment (benmoriceau) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java:167]: Done.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java
Comment (benmoriceau) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java:137]: Done

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java
Comment (benmoriceau) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java:144]: Done, I'm not sure that I fully get the jackson part.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java
Comment (tryangul) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java:146]: Nit: Comment formatting is weird here.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java
Comment (tryangul) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java:147]: This is the magic line 🪄

**Review by tryangul — APPROVED** [review]
LGTM, but will defer to destinations folks

**Pull request merged**
